src/sysext/sysext.c

```diff
@@ -0,0 +1,987 @@
+
+                case 'l':
+                        arg_action = ACTION_LIST;
+                        break;
+
+                case ARG_ROOT:
+                        r = parse_path_argument_and_warn(optarg, false, &arg_root);
+                        if (r < 0)
+                                return r;
+                        break;
+
+                case ARG_JSON:
+                        r = json_parse_cmdline_parameter_and_warn(optarg, &arg_json_format_flags);
+                        if (r <= 0)
+                                return r;
+
+                        break;
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached("Unhandled option");
+                }
+
+        if (argc - optind > 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Unexpected argument.");
+
+        return 1;
+}
+
+static int parse_env(void) {
+        _cleanup_strv_free_ char **l = NULL;
+        const char *e;
+        char **p;
+        int r;
+
+        e = secure_getenv("SYSTEMD_SYSEXT_HIERARCHIES");
+        if (!e)
+                return 0;
+
+        /* For debugging purposes it might make sense to do this for other hierarchies than /usr/ and
+         * /opt/, but let's make that a hacker/debugging feature, i.e. env var instead of cmdline
+         * switch. */
+
+        r = strv_split_full(&l, e, ":", EXTRACT_DONT_COALESCE_SEPARATORS);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse $SYSTEMD_SYSEXT_HIERARCHIES: %m");
+
+        STRV_FOREACH(p, l) {
+                if (!path_is_absolute(*p))
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Hierarchy path '%s' is not absolute, refusing.", *p);
+
+                if (!path_is_normalized(*p))
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Hierarchy path '%s' is not normalized, refusing.", *p);
+
+                if (path_equal(*p, "/"))
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Hierarchy path '%s' is the root fs, refusing.", *p);
+        }
+
+        if (strv_isempty(l))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "No hierarchies specified, refusing.");
+
+        strv_free_and_replace(arg_hierarchies, l);
+        return 0;
+}
+
+static int run(int argc, char *argv[]) {
+        _cleanup_(hashmap_freep) Hashmap *images = NULL;
+        int r;
+
+        log_show_color(true);
+        log_parse_environment();
+        log_open();
+
+        r = parse_argv(argc, argv);
+        if (r <= 0)
+                return r;
+
+        r = parse_env();
+        if (r < 0)
+                return r;
+
+        if (!arg_hierarchies) {
```

Comment:
Not a matter, but this condition is always true.

Reply:
the parse_env() called immediately before initializes arg_hierarchy from an env var, if that env var is set, so while arg_hierarchies is typically NULL when we come here, it doesn't have to be if people use that env var.